### DISCUSSION TITLE
with-sentry example doesn't work for server-side errors when deployed to Vercel

### DISCUSSION THREAD

**Discussion by georgiosd** [vercel/next.js#14758] [Help]
Insanely weird. If I build the with-sentry example locally and set the environment variables, it works fine.

If I then deploy to Vercel, the server-side tracking is not working at all.

The sourcemaps get uploaded, which should be testament to the environment being set up correctly.

Any ideas?

**Comment by TheHolyWaffle** [comment]
I managed to get a sentry setup fully working with sourcemap support for both server and client side errors. Will reply back once I'm home. 

I remember having to do quite some strange things in order to make the server side error reporting work.

**Comment by georgiosd** [comment]
That would be very much appreciated!

**Comment by JavierMartinz** [comment]
Hey @TheHolyWaffle 

Can you tell us how you did it, please? 😅

**Comment by xyy94813** [comment]
In practice, our App needs to be deployed to multiple environments in production.
For example: QA, DEV, PRE-PROD, PROD

Sentry supports issues that differentiate between environments.
All we need to do is sentry initialisation to set the configuration - `environment`.

I tried customizing the `__error` page and initializing sentry when `getInitialProps` on the `__error` page, handling errors and interfacing with Sentry.
However the automatic Static Optimize only gets the runtime config at build time.

For Case: Static A Page -> `__error` page. 
What I'm doing is completely invalid.

I need to build each environment separately because of the limitations of the runtime config in Next.js.
And the difference between them is just an environment variable.

**Comment by erik-beus** [comment]
From the example https://github.com/vercel/next.js/tree/canary/examples/with-sentry I'm only able to get 1 of 4 server side errors to appear in Sentry. I've deployed the app here: https://with-sentry-app.vercel.app/
The error that works is when the exception is manually caught: `Sentry.captureException(error)`

Previously I've also tried to follow [this blog post](https://leerob.io/blog/configuring-sentry-for-nextjs-apps) from @leerob but without luck.

Repo is here https://github.com/erik-beus/with-sentry-app

**Reply by leerob** [reply]
I'm sorry you're having issues with this. That example should be working, please file an issue with a reproduction if it's not.

P.S. We're working with the team at Sentry to improve the integration with Next.js 😄

**Reply by georgiosd** [reply]
@leerob Isn't the example a reproduction?

What's your timeline with the Sentry work?

**Reply by erik-beus** [reply]
I'll open an issue and reference my repo + deployment.

It's much appreciated if the integration can be simplified in the future. Right now it's a lot of steps to get it setup involving both the config and the _error files with a few hacks for some of the open issues.

**Reply by vvo** [reply]
> The error that works is when the exception is manually caught: `Sentry.captureException(error)`

From the with-sentry example, this seems like a documented behavior as of now:

https://github.com/vercel/next.js/blob/57ed93bc5e3eb47a9cfa71423d4a58021267905b/examples/with-sentry/pages/index.js#L21-L22

Here's what you need to do for now: https://dev.to/aryanjnyc/add-sentry-to-vercel-next-js-api-routes-85o

**Comment by vvo** [comment]
I've also followed every sentry documentation I could and came to the same conclusion than @erik-beus: the with-sentry example fails for API routes. You have to manually catch errors otherwise they are never sent to sentry.

Frontend errors and SSR errors seem to be fine. But API routes will not send any errors to Sentry. And I have no idea how to debug further.

**Reply by vvo** [reply]
A small note to myself: for API routes you need to wrap them yourself, example: https://dev.to/aryanjnyc/add-sentry-to-vercel-next-js-api-routes-85o

**Comment by georgiosd** [comment]
Looks like there is now a sentry plugin for nextjs - anyone tried it? https://www.npmjs.com/package/@sentry/nextjs

**Reply by vvo** [reply]
@georgiosd Thanks! Very curious to try it. My own Sentry implementation looks like a forgotten bowl of spaghetti.

**Comment by chris-erickson** [comment]
Has anyone had luck with this in the last year or two? Sentry seems to think there's issues at the Vercel level that might be causing these to never get sent. I can't get it to work at all, while client-side and local builds do seem to work.
